### ISSUE TITLE
undefined symbols when using multiple extension modules

### ISSUE THREAD

**Issue by ahgamut** [pybind/pybind11#2129]
I'm trying to use multiple extension modules with a project in pybind11. I followed the [snippets provided](https://pybind11.readthedocs.io/en/stable/advanced/misc.html#partitioning-code-over-multiple-extension-modules), but I get an `undefined symbol` error.

I have a package named `mypackage` with two extension modules: `mypackage.core` (containing `BaseClass`) and `mypackage.ext` (containing `SubClass`, which is derived from `BaseClass`). I use `setuptools` to build `mypackage`, and my `setup.py` is similar to the one in the [example](https://github.com/pybind/python_example). The modules compile, but I get an `ImportError` when trying to load `SubClass`:

```pycon
>>> from mypackage.core import BaseClass  # no error
>>> from mypackage.ext import SubClass
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  ImportError: /home/user/anaconda3/envs/my-environ/lib/python3.6/site-packages/mypackage/ext.cpython-36m-x86_64-linux-gnu.so: undefined symbol: _ZN10CoreStructC2Ev
```

Looking through the two `.so` modules with `nm`, I see that `_ZN10CoreStructC2Ev` refers to the constructor of `CoreStruct`/`BaseClass`, and it is undefined in `ext.so`. 

Is it a `setuptools`/linker error or is the example code incorrect? 
The package runs properly when using a single extension module.

## example files

`src/mypackage/core.h`:

```cpp
#include <pybind11/pybind11.h>
class PYBIND11_EXPORT CoreStruct {
   public:
    int id;

    CoreStruct();
    CoreStruct(int);
    int add(int, int);
};
```

`src/mypackage/core.cpp`:

```cpp
#include <core.h>

int CoreStruct::add(int x, int y) { return x + y + 1; }
CoreStruct::CoreStruct() { id = 0; } // if this is in core.h, causes pybind11::error_already_set
CoreStruct::CoreStruct(int x) { id = x; }

PYBIND11_MODULE(core, m) {
    using namespace pybind11;
    pybind11::class_<CoreStruct>(m, "BaseClass")
	.def(init<>())
	.def(init<int>())
	.def_readwrite("ID", &CoreStruct::id)
	.def("add", &CoreStruct::add)
	.def("__str__", [](CoreStruct& zz) { return "CoreStruct"; });
}
```

`src/mypackage/ext.cpp`:

```cpp
#include <core.h>

class PYBIND11_EXPORT ExtStruct : public CoreStruct {
    using CoreStruct::CoreStruct;

   public:
    float y;

    ExtStruct(int i, float y) {
	id = i;
	y = y;
    }
    ExtStruct(float y) {
	id = -1;
	y = y;
    }
    float subtract(float x, float y) { return x - y; }
};

void init_float(pybind11::module& m) {
    using namespace pybind11;
    auto core_module = pybind11::module::import("core"); // maybe import the class via attr?
    class_<ExtStruct, CoreStruct>(m, "SubClass")
	.def(init<>())
	.def(init<int, float>())
	.def(init<float>())
	.def("subtract", &ExtStruct::subtract)
	.def("__str__", [](ExtStruct& zz) { return "ExtStruct"; });
}

PYBIND11_MODULE(ext, m) { init_float(m); }
```

**Comment by bstaletic** [comment]
You have 2 options here:

1. Define `CoreStruct::CoreStruct()` inside the class.
2. Change `struct CoreStruct` to `struct PYBIND11_EXPORT CoreStruct`

If you still have troubles, come over to [gitter](https://gitter.im/pybind/Lobby).

**Comment by adrn** [comment]
@bstaletic Stumbling on this because I have a similar issue, but I don't understand your answer: Can you explain the two options you proposed a bit more? For 1., what do you mean? For 2., it looks like @ahgamut's class is defined already with `PYBIND11_EXPORT`...?

**Comment by adrn** [comment]
Connecting to #1391 in case others stumble on this thread
